Setup counter

DaengDo opened the pull request:
- 의도대로 작동하나, 관심사 분리를 하지 않은 index.jsx 파일 셋팅
- ui 요소, 상태 관리, 전체 app 으로 일단 분리 시작할 것


으아 집안일 이슈로 오늘에서야 영상을 다 봤습니다. 차근차근 그러나 늦지않게 하겠습니다!!

DaengDo committed:
Setup counter

- 의도대로 작동하나, 관심사 분리를 하지 않은 index.jsx 파일 셋팅
- ui 요소, 상태 관리, 전체 app 으로 일단 분리 시작할 것

DaengDo committed:
Seperate components

- 전체 App 요소, 카운터의 각 기능, 상수 값 분리

DaengDo commented:
## 전체 구조
`src`
├─`const` : 상수 값을 저장하기 위한 폴더입니다.
│  └─`counter.js` : 상수 값이 저장되어 있습니다. 참조로 인한 변화를 방지하기 위해 `Object.freeze()` 를 했습니다.
├─`Counter.jsx`: 카운터 프로그램입니다.
├─`CounterButton.jsx`: 카운터의 클릭 버튼 요소입니다.
├─`CounterPads.jsx`: 카운터의 숫자 패드 요소입니다.
└─`index.jsx`: 전체 App의 껍데기(?)입니다.

## 질문 사항

- 현재 `Counter.jsx` 에서 `useState`로 관리 중인 상태가 해당 컴포넌트에서는 직접적으로 쓰이지 않습니다. <BR>이 경우 `useState`로 관리중인 상태를 다른 `js` 폴더로 분리하는 것이 좋나요?
- 너무 잘개잘개 쪼개면(?) 오히려 불편할 때도 있을까요? (당장 저에겐 굉장히 합리적으로 보입니다)

### 여담

저번 1주차 때 친절하신 리뷰 사랑합니다 😍 receive object!! 특히 너무 좋았습니다.

hannut91 commented:
> 현재 Counter.jsx 에서 useState로 관리 중인 상태가 해당 컴포넌트에서는 직접적으로 쓰이지 않습니다.
이 경우 useState로 관리중인 상태를 다른 js 폴더로 분리하는 것이 좋나요?

좋은 고민입니다 ㅎㅎ 지금 단계에서는 Counter컴포넌트가 분리하시려는 그 역할을 하고 있어서 분리해도 많이 달라질 것 같지 않습니다. 처음부터 너무 분리하는 것보다는 필요한 떄에 적절히 분리해주는게 좋아보여요

hannut91 commented on src/const/counter.js:
```diff
@@ -0,0 +1,10 @@
+// ui 데이터 모음
+const counterInterface = {
+  counterLabel: 'Counter',
+  counterButtonLabel: 'Click me!',
+  counterPads: [1, 2, 3, 4, 5],
+};
```
4줄 코드에 `counter`가 4번이나 들어갔네요.

hannut91 commented on src/index.jsx:
```diff
@@ -0,0 +1,16 @@
+// 전체 App 요소
+function App() {
+  return (
+    <div>
+      <Counter />
+    </div>
```
```suggestion
    <Counter />
```

이렇게 해줘도 되겠네요

hannut91 commented on src/CounterButton.jsx:
```diff
@@ -0,0 +1,24 @@
+  );
+}
+
+export default CounterButton;
```
`export default`를 따로 쓰신 이유가 있을까요?

DaengDo commented:
## export default 를 따로 쓰신 이유가 있을까요?

>제 나름의 사용 이유
- `as`없이 이름 변경 가능
  - 상수 파일의 `counterInterface` 객체를 `export default` 해서 다른 모듈에서 상수임을 들어내기 위해 `COUNTER_CONST` 로 import 했습니다.
- `import` 를 할 때 `import React, { useState } from 'react'` 와 같이 <br>주요 export 대상과 곁가지(?) 요소들의 구분이 가능할 것 같아 보여서 `export default` 를 썼습니다.

>사실은....고민 없는 무지성 사용
- 하지만 말씀대로 `import * as ~ from 'path'` 로 원한다면 이름을 바꿀 수 있다는 점
  - 오히려 이름을 바꿔서 `import` 하면 어떤 요소인지 더 햇갈릴 수 있을지도...<br>`as`로 이름을 바꿨다는 것을 명시해주는 것이 의도를 더 잘 표현할 수 있겠네요 ㅠㅠ
- 해당 컴포넌트에서는 한 가지만 `export` 하고 있다는 점을 간과한 것 같습니다.
- 또 검색을 해보고 웹펙과 같은 번들링 툴에서 `import` 하지 않은 대상은 코드에서 훌훌 털어준다는(!) 사실을 처음 알았습니다.

## 4번 줄 코드에 counter 가 4번이나 들어갔네요

- 어...어쩌죠..? 다른 컴포넌트에서 상수를 가져다가 사용할 때 counter가 명시되어 있지 않다면 햇갈리지 않나요? 상수의 관리는 잘 모르겠습니다..😥

Commit:
Change 'export default' to 'export'

- export default 를 export 로 바꾼 후 이름 변경이 필요한 경우 as로 명시
- counter.js 파일의 경우 eslint 가 export default 권장

hannut91 commented:
```
const counter = {
  label: 'Counter',
  buttonLabel: 'Click me!',
  pads: [1, 2, 3, 4, 5],
}
```

이런 모양이 되어야 하지 않나 생각했었어요 ㅎㅎ 그럼 쓰는 입장에서는 이렇게 되겠죠

```jsx
<Button>
  {counter.label}
</Button>
```

hannut91 commented:
테스트가 통과하지 않네요. 요구사항에 맞춰서 구현해 보세요

DaengDo committed:
Pass the test

- eslint, test 통과

hannut91 commented:
테스트는 통과했지만 `eslint`에서 실패하는 군요 ㅠㅜ
